Rakefile

```diff
@@ -1,72 +1,50 @@
-require "json"
 require "pathname"
 
-CONFIG_FILE  = Pathname.new(__dir__).join(ENV["CNFFILE"] || "config.json")
 BUILD_ROOT   = Pathname.new(ENV["OBJDIR"] || "build")
 INSTALL_ROOT = Pathname.new(ENV["PREFIX"] || "/opt/ood")
+VENDOR_DIR   = Pathname.new(ENV["VENDOR_DIR"] || BUILD_ROOT.join("vendor").expand_path )
 
-def all_components
-  JSON.parse(CONFIG_FILE.read).map { |c| Component.new(c) }
-end
-
-class Component
-  attr_reader :name
-  attr_reader :url
-  attr_reader :tag
-
-  def initialize(opts = {})
-    @name = opts.fetch("name") { raise ArgumentError, "No name specified. Missing argument: name" }.to_s
-    @url  = opts.fetch("url")  { raise ArgumentError, "No url specified. Missing argument: url" }.to_s
-    @tag  = opts.fetch("tag")  { raise ArgumentError, "No tag specified. Missing argument: tag" }.to_s
-    @app  = opts.fetch("app", nil)
-  end
-
-  def app?
-    !!@app
-  end
-
-  def build_root
-    app? ? BUILD_ROOT.join("apps", name) : BUILD_ROOT.join(name)
-  end
-
-  def download_url
-    "#{url}/archive/#{tag}.tar.gz"
-  end
+task :default => :build
 
-  def file_url
-    "apps/#{name}/."
-  end
+def apps
+  %w(activejobs bc_desktop dashboard file-editor files myjobs shell).map {|name| BUILD_ROOT.join("apps", name) }
 end
 
-task :default => :build
+def build_apps
+  app_build_dir = BUILD_ROOT.join("apps")
+  rm_rf app_build_dir if app_build_dir.directory?
+  mkdir_p BUILD_ROOT unless BUILD_ROOT.directory?
+  cp_r "apps", BUILD_ROOT
 
-all_components.each do |c|
-  file c.build_root => CONFIG_FILE do
-    rm_rf c.build_root if c.build_root.directory?
-    mkdir_p c.build_root unless c.build_root.directory?
-    cp_r c.file_url, c.build_root
-    setup_path = c.build_root.join("bin", "setup")
+  apps.each do |app|
+    setup_path = app.join("bin", "setup")
     if setup_path.exist? && setup_path.executable?
-      sh "PASSENGER_APP_ENV=production PASSENGER_BASE_URI=/pun/sys/#{c.name} #{setup_path}"
+      args = "PASSENGER_APP_ENV=production PASSENGER_BASE_URI=/pun/sys/#{app.basename}"
+      args = args + " BUNDLE_DIR=#{VENDOR_DIR.join("bundle")}"
+
+      sh "#{args} #{setup_path}"
     end
-    c.build_root.join("VERSION").write(c.tag) if c.app?
   end
+
 end
 
 def proxy_components
   %w(ood-portal-generator mod_ood_proxy ood_auth_map nginx_stage).map {|name| BUILD_ROOT.join(name) }
 end
 
-proxy_components.each do |build_root|
-  file build_root => 'nginx_stage/lib/nginx_stage/version.rb' do
```

Comment:
This bit stopped working and I'm not sure how to get it back or what it does, so I changed it to a more straight forward function.

Reply:
What it does is useful to understand: for each proxy component, create a take task. “file” is a special type of rake task where the “dependency” is not another task but a file. This has interesting utility in some cases.

Regardless, the end result is a rake task for building each individual app and infrastructure component, and then the build task that depends on all these tasks, meaning when you execute build it executes all of the dependent tasks:

```
task :build => all_components.map(&:build_root) + proxy_components
```

>An important part of the file task is that it's not run unless you need to run it. The build system looks at the files and only runs the task if the output file does not exist or its modification date is earlier than the input file.

https://martinfowler.com/articles/rake.html

Reply:
BTW that Martin Fowler link is a great overview of Rake which totally rocks and one very useful and simple feature is if you execute rake in a subdirectory of your app by default it will execute the rake task in the Rakefile from the root directory of your app as though you cd-d to that directory first - just like how git works. Very useful functionality.